### PR TITLE
Specific message for unauthorized errors

### PR DESCRIPTION
**Pull request by Wauplin**
Currently any HTTP 401 Unauthorized is processed as "Repository not found" with an error message quite unhelpful for gated repos. This PR mentions the fact that you must authenticate if a repo is gated.

See https://github.com/huggingface/moon-landing/issues/3903#issuecomment-1320137948 (internal link) and [this gist](https://gist.github.com/Wauplin/46c27ad266b15998ce56a6603796f0b9).

(cc @julien-c: realized that when making a local test for https://github.com/huggingface/huggingface_hub/issues/1198)

### CODE REVIEW

File: src/huggingface_hub/utils/_errors.py [hunk 1]
```diff
@@ -257,13 +257,18 @@ def hf_raise_for_status(
             raise EntryNotFoundError(message, response) from e
 
         elif error_code == "RepoNotFound" or response.status_code == 401:
+            # 401 is misleading as it is returned for:
+            #    - private and gated repos if user is not authenticated
+            #    - missing repos
+            # => for now, we process them as `RepoNotFound` anyway.
+            # See https://gist.github.com/Wauplin/46c27ad266b15998ce56a6603796f0b9
```
Comment (julien-c) [inline src/huggingface_hub/utils/_errors.py:264]: @huggingface/moon-landing-back to check the accuracy of this comment

**Review by julien-c — COMMENTED** [review]
Ping some one from Hub backend team to check the comment.

Other than that, LGTM

File: src/huggingface_hub/utils/_errors.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/utils/_errors.py:264]: Very much related to this comment: https://github.com/huggingface/moon-landing/issues/3903#issuecomment-1320137948

File: src/huggingface_hub/utils/_errors.py [hunk 2]
```diff
@@ -257,13 +257,18 @@ def hf_raise_for_status(
...
             message = (
                 f"{response.status_code} Client Error."
                 + "\n\n"
                 + f"Repository Not Found for url: {response.url}."
                 + "\nPlease make sure you specified the correct `repo_id` and"
-                " `repo_type`."
-                + "\nIf the repo is private, make sure you are authenticated."
+                " `repo_type`.\nIf you try to access a private or gated repo, make"
```
Comment (osanseviero) [inline src/huggingface_hub/utils/_errors.py:270]:
```suggestion
                " `repo_type`.\nIf you are trying to access a private or gated repo, make"
```

File: tests/test_hf_api.py [hunk 3]
```diff
@@ -954,10 +954,10 @@ def test_create_commit_repo_does_not_exist(self) -> None:
                     " Found for url:"
                     f" {self._api.endpoint}/api/models/{USER}/repo_that_do_not_exist/{route}/main.\nPlease"
                     " make sure you specified the correct `repo_id` and"
-                    " `repo_type`.\nIf the repo is private, make sure you are"
-                    " authenticated.\nNote: Creating a commit assumes that the repo"
-                    " already exists on the Huggingface Hub. Please use `create_repo`"
-                    " if it's not the case."
+                    " `repo_type`.\nIf you try to access a private or gated repo, make"
```
Comment (osanseviero) [inline tests/test_hf_api.py:957]:
```suggestion
                    " `repo_type`.\nIf you are trying to access a private or gated repo, make"
```

**Review by osanseviero — APPROVED** [review]
LGTM as well! Let's wait for some backend folks too!

**Review by coyotte508 — APPROVED** [review]

**Pull request merged**